### DISCUSSION TITLE
How to set different refresh token expiration times for multiple clients under the same realm？

### DISCUSSION THREAD

**Discussion by Jenxassy** [keycloak/keycloak#10183] [Q&A] [answered]
hello,

I am confused about setting the refresh token expiration time on the client.

I found two parameters ssoSessionMaxLifespan and ssoSessionIdleTimeout in the code, it seems that the refresh token expiration time of each client must be shorter than these two times.

https://github.com/keycloak/keycloak/blob/61ea2dd1bb983cbc6a37f82c8e59d5e1b7999c00/services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java#L1082

```
private int getRefreshExpiration() {
    int sessionExpires = userSession.getStarted()
                    + (userSession.isRememberMe() && realm.getSsoSessionMaxLifespanRememberMe() > 0
                        ? realm.getSsoSessionMaxLifespanRememberMe()
                        : realm.getSsoSessionMaxLifespan());
    
    ……
    int expiration = Time.currentTime() + (userSession.isRememberMe() && realm.getSsoSessionIdleTimeoutRememberMe() > 0
                    ? realm.getSsoSessionIdleTimeoutRememberMe()
                    : realm.getSsoSessionIdleTimeout());
    
    ……
    return expiration <= sessionExpires ? expiration : sessionExpires;
}
```

If I need some clients to specify the expiration time of the refresh token, it may be longer than the realm-level ssoSessionIdleTimeout, while other clients need to use the default value ssoSessionIdleTimeout? 

At present, the best way I can find is to modify the ssoSessionIdleTimeout and ssoSessionMaxLifespan at the realm level, and then set ssoSessionIdleTimeout at each client level to control the refresh token expiration time.

Is there a way to modify only the parameters of the client without modifying the ssoSessionIdleTimeout at the realm level to achieve this purpose? Or increase ssoSessionMaxLifespan and make ssoSessionIdleTimeout the default value of each client.

Because it is possible that the refresh time required by the client is longer than the ssoSessionIdleTimeout time at the realm level again.

Many thanks to everyone who noticed.

**Comment by Jenxassy** [comment, marked as answer]
I think I found a way to convert the generated refresh_token, thank you all.

**Reply by fawadasaurus** [reply]
How?
